[code-quality/MM-45507]: Badge component

michelengelen opened the pull request:
#### Summary
the `Badge` component was cluttery and not optimally aligning with the design.

this PR fixes that and also renames it to `Tag` which is the term used for it on the design files as well (previously it got confused with the a badge, the mention- and unread-indicator, so this change seems appropriate.

there were also some slight changes made to the component itself:
- removed the `show` prop and instead added conditional rendering to where it got used
- added a `capitalize` prop to capitalize the text inside
- changed the structure so that the component does not allow children anymore, but instead uses a `text` prop
- added a `icon` prop (valid values are the icon names, `IconGlyphTypes`,  from the compass-icon package and gets the Icon component from the `glyphMap`), so passing in an icon name will render the icon as a leading decorator
- added a `size` prop with the options `'xs'` (default) and `'sm'`

the only "real" visible change is with the `BotBadge` (now `BotTag`) and `GuestBagde` (now `GuestTag`), since those had only a border radius of 2px, but now align with the `Tag` components 4px radius

#### Ticket Link
[MM-45507](https://mattermost.atlassian.net/browse/MM-45507)

#### Related Pull Requests
n/a

#### Screenshots
n/a

#### Release Note
```release-note
NONE
```


[MM-45507]: https://mattermost.atlassian.net/browse/MM-45507?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/238758

matthewbirtch reviewed (changes requested):
@michelengelen thanks for this. I went through as much as I could, but I'm not really sure if I caught all places this change applies. 

I noticed a number of things that I think need tweaking:

1. The tag seems to span wider than it should in some cases. This is how the `Tag` component is displaying in the Top Threads widget and the top threads modal within Insights
<img width="859" alt="image" src="https://user-images.githubusercontent.com/2040554/189197973-78ecbf8e-9827-4d72-9462-7d1cb17a4ea7.png">
<img width="436" alt="image" src="https://user-images.githubusercontent.com/2040554/189197996-2f58c79b-15bd-4069-8302-3cbcbc621ac1.png">

2. The tag doesn't look like the right height in this system console context. The height is 24px, but I think it's supposed to be 16 if it's the small version (I'm assuming it's the small version because of the font-size that is used, but I'm not 100% sure what it's supposed to be). I can't quite figure out what's causing it to have the extra height though.

<img width="1037" alt="image" src="https://user-images.githubusercontent.com/2040554/189204432-556ce60a-0cec-49cc-bb9f-dbf213dadc1f.png">

3. Can you verify where we are getting the color used on the new tag here? Is that a hard-coded color?
<img width="763" alt="image" src="https://user-images.githubusercontent.com/2040554/189205699-ae306940-ee96-4631-bc50-12a55b7dead2.png">

4. The height of the tag that shows in the info RHS for DMs also doesn't look right.
<img width="400" alt="image" src="https://user-images.githubusercontent.com/2040554/189206038-bb9a76a9-11b4-4b52-abcf-505d0435f969.png">

5. The Forward `NEW` tag isn't displaying as the right color. It's just displaying as the standard default tag color. We had it green initially (to match the tour point pulsing dot color)
<img width="352" alt="image" src="https://user-images.githubusercontent.com/2040554/189206486-72856c9d-bdad-40c4-96b3-6d7e8ba6e987.png">

6. Noticed the Bot tag doesn't show correctly in the channel intro
<img width="571" alt="image" src="https://user-images.githubusercontent.com/2040554/189206749-a79bf33b-6e81-4d26-870c-20b3548ae9d3.png">
<img width="626" alt="image" src="https://user-images.githubusercontent.com/2040554/189206907-46e26e55-b443-4638-b81b-423f728dbc08.png">

7. Somehow it seems that BOT tags are showing for all messages in the search results RHS when they are clearly not bots.
<img width="399" alt="image" src="https://user-images.githubusercontent.com/2040554/189207747-aa984208-6f0b-4d48-95f8-1948c92bfbb1.png">

8. It seems the `BOT` and `GUEST` tag are showing for all items in the @mention autocomplete box as well. 
<img width="559" alt="image" src="https://user-images.githubusercontent.com/2040554/189207999-c096427b-68e7-4408-8676-4a58f1f5d47c.png">

9. Same thing seems to be happening for the Channel Switcher as well
<img width="697" alt="image" src="https://user-images.githubusercontent.com/2040554/189208299-a7858bf6-9232-4225-8253-00ac129eb923.png">

10. And the `BOT` tag is always showing in the search from:user autocompelete
<img width="454" alt="image" src="https://user-images.githubusercontent.com/2040554/189212274-94c1712f-7917-4438-8bd5-bd3b3e2aae23.png">

michelengelen commented:
@matthewbirtch

>     1. The tag seems to span wider than it should in some cases. This is how the `Tag` component is displaying in the Top Threads widget and the top threads modal within Insights

Fixed ... it was an override from outside that did not play well here.

>     2. The tag doesn't look like the right height in this system console context. The height is 24px, but I think it's supposed to be 16 if it's the small version (I'm assuming it's the small version because of the font-size that is used, but I'm not 100% sure what it's supposed to be). I can't quite figure out what's causing it to have the extra height though.

Fixed

>     3. Can you verify where we are getting the color used on the new tag here? Is that a hard-coded color?

Fixed ... yse, that was a hard-coded color. I changed that to use the semantic colors now.
 
>     4. The height of the tag that shows in the info RHS for DMs also doesn't look right.

Fixed ... same as 2.

>     5. The Forward `NEW` tag isn't displaying as the right color. It's just displaying as the standard default tag color. We had it green initially (to match the tour point pulsing dot color)

Fixed ... I was missing an opacity property here.

>     6. Noticed the Bot tag doesn't show correctly in the channel intro

Fixed ... same as 2.

>     7. Somehow it seems that BOT tags are showing for all messages in the search results RHS when they are clearly not bots.

Fixed ... they were still using the old `show` prop. I missed that since it did not get caught by typescript

>     8. It seems the `BOT` and `GUEST` tag are showing for all items in the @mention autocomplete box as well.

Fixed ... same as 7.

>     9. Same thing seems to be happening for the Channel Switcher as well

Fixed ... same as 7.

>     10. And the `BOT` tag is always showing in the search from:user autocompelete

Fixed ... same as 7.

I am currently going over every override and check if that is really necessary. I will notify you once I have finished that.

koox00 commented:
@michelengelen I had already made some changes, to conform to our design, on the Badge component (for the Post Proirity feature). I have merged the PR just now.
Sorry about that.

michelengelen commented:
@koox00 no worries ... i will incorporate them into mine!

koox00 commented:
Please when you do set this env var `MM_FEATUREFLAGS_POSTPRIORITY=true` so you can also see/review the PostPriority badges with your changes :)
`

michelengelen commented:
@matthewbirtch this PR is now ready for a new round of reviews

matthewbirtch commented:
> @matthewbirtch and @koox00 this PR is now ready for a new round of reviews

As we discussed in our chat there are a few locations still that need the margin added between the username and tag. Once those are fixed, we're good to go

matthewbirtch reviewed (approved):
I think we're good to go @michelengelen. Thanks!

michelengelen commented:
@michaelkochub

michelengelen commented:
@mickmister ☝🏼

koox00 commented:
It would be preferable to use `git mv` when renaming files, so we could keep history of the files. I have no idea if it's reversible right now, just saying for the future!

michelengelen commented:
> It would be preferable to use `git mv` when renaming files, so we could keep history of the files. I have no idea if it's reversible right now, just saying for the future!

Well, thanks for the tip @koox00 ... I did not know that this is different from just renaming files ... Will try to remember that and use it for future changes. Thanks!

pvev commented:
thanks @michelengelen everything looks good so far, I only encountered an issue in the invite people modal in the cloud version for the tag used to notify about the premium feature of guest invite, see attached:

<img width="965" alt="Captura de Pantalla 2022-09-28 a las 16 03 03" src="https://user-images.githubusercontent.com/10082627/192841481-d76eb86d-d202-4ccc-9f51-c952ee3f7bd2.png">

That can be fixed just by removing the property `display: flex` in the `invite_as.scss > &__restricted-indicator-badge` as shown in the picture:

<img width="1566" alt="Captura de Pantalla 2022-09-28 a las 16 14 42" src="https://user-images.githubusercontent.com/10082627/192841536-da245c78-b71a-4320-bfac-649dfb98dac3.png">

other thing I found was the bot tag not having margin with the text in the direct message modal:
<img width="627" alt="Captura de Pantalla 2022-09-28 a las 16 19 31" src="https://user-images.githubusercontent.com/10082627/192841873-6ef80b59-48b1-4433-a97d-d2760eb127f0.png">

Besides this, it should be good to go 👍

M-ZubairAhmed commented:
@michelengelen please let me know when its ready i will start the review :)

michelengelen commented:
> @michelengelen please let me know when its ready i will start the review :)

this is ready now @M-ZubairAhmed

michelengelen commented on components/admin_console/system_roles/system_role/system_role_users/system_role_users.tsx:
```diff
@@ -174,22 +175,19 @@ export default class SystemRoleUsers extends React.PureComponent<Props, State> {
+                    new: usersToAdd[user.id] ? (
+                        <Tag
+                            variant={'info'}
+                            uppercase={true}
```
I am not sure what you mean with that ... could you elaborate?

M-ZubairAhmed replied:
```
<Tag
   uppercase
   variant={'info'}
.
.
.
```

M-ZubairAhmed replied:
oh i wasnt aware we cannot due to eslint rule

michelengelen replied:
I will check that, but there was a rule in place preventing shortened Boolean props

michelengelen commented on components/widgets/tag/tag.tsx:
```diff
@@ -0,0 +1,172 @@
+    background: rgba(var(--tag-bg), var(--tag-bg-opacity));
+    color: rgb(var(--tag-color));
+
+    ${({onClick}) => typeof onClick === 'function' && (
```
its just a safety measure, since short circuit will return the value if it is falsy, so for instance when `onClick` is `0` it will just return `0` and might potentially break. This is not the case for boolean values

michelengelen commented on components/channel_info_rhs/about_area_dm.tsx:
```diff
@@ -5,12 +5,14 @@ import React from 'react';
 import {useIntl} from 'react-intl';
 
+import BotTag from 'components/widgets/tag/bot_tag';
+
```
ok, changed the order and grouping according to the new eslint rules

michelengelen commented on plugins/export.js:
```diff
@@ -63,7 +64,7 @@ window.Components = {
     Avatar,
     imageURLForUser,
-    BotBadge,
+    BotBadge: BotTag,
```
But the change here is minimal. IMO this can stay as is, but we should have our eyes open for any upcoming bug tickets.

The API for the component did not change, so it can just be used as-is

michelengelen commented:
@jgilliam17 this is ready for a review now

michelengelen commented:
@jgilliam17 kind reminder to review

jgilliam17 commented:
Thanks @michelengelen apologies for the delay 
I'm not seeing my comment from Friday here, adding screen shots again
Oh, and can you please fix the conflict so I can trigger e2es.

invite people modal shows a small glitch with the tag
Other then that, manually tested - looks good 

<img width="637" alt="Screen Shot 2022-10-14 at 11 03 14 AM" src="https://user-images.githubusercontent.com/52937121/196202563-6a520445-0f47-4488-9437-03c9c6c56f80.png">
<img width="660" alt="Screen Shot 2022-10-14 at 11 02 50 AM" src="https://user-images.githubusercontent.com/52937121/196202566-3a484c48-fb41-40a8-a6fe-408a997a5d85.png">

jgilliam17 commented:
Thanks @michelengelen 
Professional feature indicator positioning on the invite modal is fixed. I'm running e2es and will post the report soon.
Seeing 2 issues
1. Guest label on channel members list is missing spacing
<img width="543" alt="Screen Shot 2022-11-09 at 10 13 05 AM" src="https://user-images.githubusercontent.com/52937121/200869322-1e7b7ae8-ab9c-48d1-b9c0-205fdbb948ff.png">

2. Forward `new` tag seems to have reversed function. It is not present on fresh install, shows up after user forwards their first post and persist on future forward post actions.

jgilliam17 commented:
@michelengelen Can you please resolve conflicts and update before I trigger e2es? thanks 🙂

koox00 commented on components/widgets/tag/__snapshots__/tag.test.tsx.snap:
```diff
@@ -0,0 +1,89 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`components/widgets/tag/Tag should match the snapshot on show 1`] = `
```
Snapshots when we shallow mount feel a bit incomplete to me.
I'd like to also assert on dom structure etc.
can we use mount for snapshot testing, since it is a "dump" component it won't be complex (3/5)

michelengelen replied:
well, i did not change the way this is tested. Would you mind opening a follow-up ticket? This could be a great HW candidate!

michelengelen commented on components/activity_and_insights/insights/top_threads/top_threads_table/top_threads_table.tsx:
```diff
@@ -218,9 +219,7 @@ const TopThreadsTable = (props: Props) => {
                                     <span className='display-name'>{displayUsername(thread.user_information as UserProfile, teammateNameDisplaySetting)}</span>
-                                    <Badge>
-                                        {thread.channel_display_name}
-                                    </Badge>
+                                    <Tag text={thread.channel_display_name}/>
```
I am not using children to limit it to strings only. Just to prevent usages it is not intended for

michelengelen commented on components/post_view/post_info/post_info.tsx:
```diff
@@ -477,10 +477,7 @@ export default class PostInfo extends React.PureComponent<Props, State> {
-        if (post.props?.priority && isPostPriorityEnabled) {
-            priority = <span className='d-flex mr-2 ml-1'><PriorityLabel priority={post.props.priority}/></span>;
-        }
+        const priority = post.props?.priority && isPostPriorityEnabled ? <PriorityLabel priority={post.props.priority}/> : null;
```
yes, since I changed the parent styling to use `display: flex;` in conjunction with `gap` the set of margins is not needed anymore

michelengelen commented:
@jgilliam17 It is ready for another review.

Will start e2e's now

michelengelen commented:
/e2e-test

jgilliam17 commented:
I need to rerun e2es, cycle failed yesterday https://automation-dashboard.vercel.app/cycles/3e177ad8-76f3-440f-a25f-dec14aa5e5bc
@michelengelen fix the conflict first? thanks 🙂

michelengelen commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/265755

michelengelen commented:
@jgilliam17 i merged master and restarted the cycle

jgilliam17 commented:
/update-branch

mattermod commented:
Error trying to update the PR.
Please do it manually.

michelengelen commented:
/e2e-test

mattermod commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/271563) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/271563)

mattermod commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/276163) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/276163)

jgilliam17 reviewed (approved):
Thanks for fixing the e2es @michelengelen 
[E2E report ](https://mattermost-cypress-report.s3.amazonaws.com/276163-05ee531-onprem-ent-webapp-pr-11100-mm-ee-test:05ee531/mochawesome.html)looks good, no PR related failures.
There are some conflicts and pending reviews, I won't merge at this time.

michelengelen commented:
@koox00 could you add your review please so we can merge this? Thanks!

koox00 commented:
I'm also noticing this in compact mode:

<img width="462" alt="Screenshot 2022-12-20 at 2 20 48 PM" src="https://user-images.githubusercontent.com/3829551/208666700-48fc7477-32fb-4250-9e4c-bfa62bd0881c.png">

notice the double colon after the username

mattermost-build commented:
E2E tests not automatically triggered, because the PR is not in a mergeable state. Please update the branch with the base branch and resolve outstanding conflicts.

michelengelen commented:
@koox00 this is ready for another round

koox00 reviewed (approved):
💯
